### DISCUSSION TITLE
tRPC with React Testing Library

### DISCUSSION THREAD

**Discussion by mehrabmp** [trpc/trpc#2773] [Q&A]
Hi there,

So i've been trying to test my components that relies on tRPC like this:
```javascript
export const Component = () => {
  const { data } = trpc.useQuery(['hello']);
  if (data === undefined) {
    return <div>Loading...</div>;
  }
  return <div>{data.message}</div>;
};
```
so when i write test like
```javascript
test('should render component', () => {
  render(<Component />);
  const element = screen.getByRole('textbox');
  expect(element).toBeInTheDocument();
});
```
it doesn't work so i think we should setup provider like `withTRCP` HOC that used for application, to provide trpc for testing enviroment.

**Comment by Railly** [comment]
You're right, one way to handle RTL with tRPC is to use a `withTRPC` HOC
```typescript
// src/utils/trpc.ts
export const trpcReact = createTRPCReact<AppRouter>(trpcConfig);
```
Then I use tRPCReact to create a provider/wrapper:

```typescript
// src/vitest/decorator.tsx
export const withTRPC = ({ children }: PropsWithChildren<{}>) => (
  <trpcReact.Provider client={trpcClient} queryClient={queryClient}>
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  </trpcReact.Provider>
);
```
Finally, to test a component:

```typescript
// src/__tests__/pages/home.spec.tsx
import { test } from "vitest";
import { render } from "@testing-library/react";
import Home from "~/pages";
import { withTRPC } from "vitest/decorator";

test("Home", async () => {
  render(<Home />, { wrapper: withTRPC });
});
```
Additionally, if you would like to get started with Next.js + RTL + tRPC, check [my answer on stackoverflow](https://stackoverflow.com/questions/73000884/testing-a-component-in-next-js-with-testing-library-that-relies-on-trcp/75688736#75688736)

**Reply by hbzhou** [reply]
got compile error can't find namepce trpcReact

**Reply by hbzhou** [reply]
fixed it out, the filename should be *.tsx

**Reply by Railly** [reply]
great, just added file names as a comment for each code block.
